File: programmers/stack_12909.java
// https://school.programmers.co.kr/learn/courses/30/lessons/12909?language=java

import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;
        Deque<Character> stack = new ArrayDeque<>();
        
        for (char c : s.toCharArray()) {
            if(c == '(') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) return false;
                stack.pop();
            }
        }
        
        return stack.isEmpty();
    }
}

File: programmers/stack_12909.py
# https://school.programmers.co.kr/learn/courses/30/lessons/12909?language=python3

def solution(s):
    st = list();
    
    for c in s:
        if c == '(' :
            st.append(c)
            
        if c == ')' :
            if len(st) < 1 :
                return False
            else :
                st.pop()
    
    return len(st) == 0
            
# 아래는 테스트로 출력해 보기 위한 코드
print( solution("(hello)()"))
print( solution(")("))       
            
